src/codegen/events/substrate.rs

```diff
@@ -32,68 +45,121 @@ impl EventEmitter for SubstrateEventEmitter<'_> {
+        // This is static and can be calculated at compile time.
+        if !event.anonymous {
+            let mut encoded =
+                format!("{}::{}", &self.ns.contracts[contract_no].name, &event.name).encode();
+            encoded[0] = 0; // Set the prefix (there is no prefix for the event topic)
+            topics.push(Expression::AllocDynamicBytes(
+                loc,
+                Type::Slice(Type::Uint(8).into()),
+                hash_len(),
+                Some(topic_hash(&encoded[..])),
+            ));
+            topic_tys.push(Type::DynamicBytes);
+        };
+
+        // Topic prefixes are static and can be calculated at compile time.
+        let mut topic_prefixes: Vec<Vec<u8>> = event
+            .fields
+            .iter()
+            .filter(|field| field.indexed)
+            .map(|field| {
+                format!(
+                    "{}::{}::{}",
+                    &self.ns.contracts[contract_no].name,
+                    &event.name,
+                    &field.name_as_str()
+                )
+                .into_bytes()
+                .encode()
+            })
+            .collect();
+
+        for (ast_exp, field) in self.args.iter().zip(event.fields.iter()) {
+            let value_exp = expression(ast_exp, cfg, contract_no, Some(func), self.ns, vartab, opt);
+            data_tys.push(value_exp.ty());
+            data.push(value_exp);
+
+            if !field.indexed {
+                continue;
             }
+
+            let value_exp = expression(ast_exp, cfg, contract_no, Some(func), self.ns, vartab, opt);
+            let encoded = Expression::AbiEncode {
+                loc,
+                tys: vec![value_exp.ty()],
+                packed: vec![],
+                args: vec![value_exp],
+            };
+            let concatenated = Expression::StringConcat(
+                loc,
+                Type::DynamicBytes,
+                StringLocation::CompileTime(topic_prefixes.remove(0)), // TODO not efficient
+                StringLocation::RunTime(encoded.into()),
+            );
+
+            vartab.new_dirty_tracker();
+            let var_buffer = vartab.temp_anonymous(&Type::DynamicBytes);
+            cfg.add(
+                vartab,
+                Instr::Set {
+                    loc,
+                    res: var_buffer,
+                    expr: concatenated,
+                },
+            );
+            let buffer = Expression::Variable(loc, Type::DynamicBytes, var_buffer);
+            let compare = Expression::UnsignedMore(
+                loc,
+                Expression::Builtin(
+                    loc,
+                    vec![Type::Uint(32)],
+                    Builtin::ArrayLength,
+                    vec![buffer.clone()],
+                )
+                .into(),
+                hash_len(),
+            );
+
+            let bigger = cfg.new_basic_block("bigger".into());
+            let done = cfg.new_basic_block("done".into());
```

Comment:
I called it `bigger`, I know. However, it's not a great name. How about `needs_hashing` or `hash_topic` or something like that.